lexers/hare.lua

```diff
@@ -14,47 +14,57 @@ lex:add_rule('whitespace', token(lexer.WHITESPACE, lexer.space^1))
 
 -- Keywords.
 lex:add_rule('keyword', token(lexer.KEYWORD, word_match{
-  'as', 'break', 'case', 'const', 'continue', 'def', 'defer', 'else', 'export', 'false', 'fn',
-  'for', 'if', 'is', 'let', 'match', 'null', 'nullable', 'return', 'static', 'struct', 'switch',
-  'true', 'type', 'use', 'yield'
+  'as', 'break', 'case', 'const', 'continue', 'def', 'defer', 'else', 'export', 'false', 'fn', 'for',
+  'if', 'is', 'let', 'match', 'null', 'nullable', 'return', 'static', 'struct', 'switch', 'true',
+  'type', 'use', 'yield'
 }))
 
 -- Functions.
+local size_builtin = 'size' * lexer.space^0 * #P('(')
 lex:add_rule('function', token(lexer.FUNCTION, word_match{
-  'len', 'alloc', 'free', 'assert', 'abort', 'size', 'append', 'insert', 'delete', 'vastart',
-  'vaarg', 'vaend'
-}))
+  'abort', 'align', 'alloc', 'append', 'assert', 'delete', 'free', 'insert', 'len', 'offset',
+  'vaarg', 'vaend', 'vastart'
+} + size_builtin))
 
 -- Types.
 lex:add_rule('type', token(lexer.TYPE, word_match{
-  'bool', 'enum', 'f32', 'f64', 'i16', 'i32', 'i64', 'i8', 'int', 'u16', 'u32', 'u64', 'u8', 'uint',
-  'uintptr', 'union', 'void', 'rune', 'str', 'char'
+  'bool', 'enum', 'f32', 'f64', 'i16', 'i32', 'i64', 'i8', 'int', 'rune', 'size', 'str', 'u16',
+  'u32', 'u64', 'u8', 'uint', 'uintptr', 'union', 'valist', 'void',
 }))
 
 -- Identifiers.
 lex:add_rule('identifier', token(lexer.IDENTIFIER, lexer.word))
 
 -- Strings.
+local sq_str = lexer.range("'", true)
 local dq_str = lexer.range('"')
 local raw_str = lexer.range('`')
-lex:add_rule('string', token(lexer.STRING, dq_str + raw_str))
+lex:add_rule('string', token(lexer.STRING, sq_str + dq_str + raw_str))
 
 -- Comments.
 lex:add_rule('comment', token(lexer.COMMENT, lexer.to_eol('//')))
 
 -- Numbers.
-lex:add_rule('number', token(lexer.NUMBER, lexer.number))
+local literal_suffix = word_match{
+	"i", "u", "z", "i8", "i16", "i32", "i64", "u8", "u16", "u32", "u64", "f32", "f64",
+}
+local bin_num = '0b' * R('01')^1
+local oct_num = '0o' * R('07')^1
```

Comment:
Is an upper-case 'o' allowed here? Also, you might want to tack on a `* -lexer.xdigit` suffix, or non-octal numbers will match immediately after this, resulting in an apparently incorrect highlight.

Reply:
> Is an upper-case 'o' allowed here?

No, just like 'b' and 'x'

> Also, you might want to tack on a `* -lexer.xdigit` suffix, or non-octal numbers will match immediately after this, resulting in an apparently incorrect highlight.

Thanks for the suggestion, fixed it

Reply:
> Is an upper-case 'o' allowed here?

No, just like 'b' and 'x'

> Also, you might want to tack on a `* -lexer.xdigit` suffix, or non-octal numbers will match immediately after this, resulting in an apparently incorrect highlight.

Thanks for the suggestion, fixed it